exec/vector/src/main/java/org/apache/drill/exec/vector/accessor/reader/UnionReaderImpl.java

```diff
@@ -268,4 +274,51 @@ public String getAsString() {
     }
     return requireReader(type).getAsString();
   }
+
+  private UnionReaderImpl getNullReader() {
+    AbstractObjectReader[] nullVariants = new AbstractObjectReader[variants.length];
+    for (int i = 0; i < variants.length; i++) {
+      nullVariants[i] = variants[i].createNullReader();
+    }
+    return new NullUnionReader(schema(), unionAccessor, nullVariants);
+  }
+
+  private static class NullUnionReader extends UnionReaderImpl {
+
+    private NullUnionReader(ColumnMetadata metadata, VectorAccessor va, AbstractObjectReader[] variants) {
```

Comment:
This approach is a bit better than the big null reader we had before. I wonder, however, why we need such a reader? Null (AKA "dummy") readers exist, but generally for specialized cases (such as a List that has no type.) I wonder, is this for the case where the value of a key/value pair is null?

If so, then two thoughts. First, the right way to handle that is to:

* Use the `isNull()` method on the `ColumnReader` class:

```
if (myDictReader.value().isNull()) {
  // Handle null value
} else {
  int x = myDictReader.value().scalar().getInt();
}
```

The idea is that null values are null: they have no structure. The approach here seems to say that null values have structure (such as a null union), so we can ask what types the null has and so on. But, this is, perhaps, too much of a fantasy.

The other suggestion would be to suggest that values cannot be null, especially if they are of structured types.

Also, note that, for unions, we already have multiple levels of nulls: The union itself can be null. When used in a list, the list entry can be null. The union can have nullable types, so that the union might have an int, but the int is null. This is already a mess, we probably don't want to make it even more complex.

Reply:
Yes, this is for the case "where the `value` of a `key`/`value` pair is null". This `value` can be `null` for a given `key` if the `key` is not present in the dict for a given row (e.g. if you have a `Map<String, Integer>` with entry `"a" : 1` and you try to `Map#get("b")`, with `Map` being a Java map in the example). That is, the value may be non-nullable, but when you fetch a `value` by `key` and there is no entry for the given key (in a row) then this `Null*Reader` is used. The `Null*Reader`s are created in https://github.com/apache/drill/pull/1870/files#diff-da401fee5976aec49192b5fb6e22de3eR83. This structures are handled outside (of particular `DictReader`) so the structure is retained for uniformity. 
The "dummy" readers are not projected (merely a placeholder, aren't they?), so this is not the way to use them(?).

Reply:
We may want to define our semantics here. If I understand your description, we have:

* The entire `DICT` is `NULL` (is this allowed)?
* The `DICT` has an entry for key `x`, but the value is null.
* The `DICT` has an entry for the `NULL` key.
* The `DICT` has no entry for the key `x`, meaning that the value is implicitly `NULL`.

We have similar situations with the `UNION` type, perhaps we can learn from that:

* The entire `UNION` can be `NULL`.
* The `UNION` can have an entry for, say, `VARCHAR`, but the `VARCHAR` is `NULL`.
* The `UNION` does not have an entry for a given type (say, `INT`), so the value is implicitly `NULL`. (Although we don't allow access to such a type.)

My suggestion would be that we use the `isNull()` method for all the `NULL` cases:

* If the `DICT` is `NULL` for a row, then `isNull()` returns true for the `DICT` reader.
* If the `DICT` has a `NULL` key, then the key `isNull()` for the key returns true.
* If the `DICT` has a key, but the value is `NULL`, then `isNull()` for the value returns true.

Not the tricky case:

* If the `DICT` does not have a key, then we may want a `hasKey()` method, and the value reader should be a Java `null` because it does not exist.

Or:

* If the `DICT` does not have a key, then the value reader's `isNull()` returns true. (This means we cannot differentiate the has-key-with-null-value vs. does-not-have-a-key cases.)

Reply:
Regarding points of your suggestion:
- `DictReaderImpl` has `NullStateReaders.REQUIRED_STATE_READER` `NullStateReader`, i.e. it can't be `NULL` (like `MapReader`).
- `DICT` can't have `NULL` key as it is `REQUIRED`.
- If the `DICT` has a key, but the value is `NULL`, then `isNull()` for the value returns `true` - yes.
- in the (tricky) case of a missing key from the `DICT`, the latter approach was taken - there is no differentiation between missing key and `NULL` value. (The new `Null*Reader`s are introduced to be used in the case of a missing key - their `isNull()` always returns `true`; note, that in case when a key is present and the value is `NULL`, current (non-null) readers' `isNull()` will return `true` for that particular position, as corresponding value reader keeps track of nullability). The former approach was discarded to ensure uniform behavior without a chance of getting `NPE` or making `null`-checks.

Reply:
@KazydubB, thanks for the explanation. It clarifies something I wondered about: the role of the null state readers. Let's talk a bit about the intended semantics for the readers.

I should be able to write code that caches the readers:

```
DictReader dr = rowReader.dict("myDict");
ScalarWriter kw = dr.key().scalar();
MapWriter vr = dr.value.tuple();
```

That is, I should be able to use the same key reader (`kr`) and value reader (`vr`) for all members. Then, I should either be able to iterate over the values:

```
while (dr.next()) {
  System.out.println(kr.getString());
  System.out.println(vr.getAsString());
}
```

If things work that way, then we don't need a special reader for the null state: the value reader (here, `vr`) would simply return `true` from `isNull()`:

```
while (dr.next()) {
  System.out.println(kr.getString());
  if (vr.isNull()) {
    System.out.println("NULL");
  } else {
    System.out.println(vr.getAsString());
  }
}
```

(Ignoring, for the moment, that `getAsString()` already does a null check.)

Note that, if we read past the end, both the key and value writers will point to nothing. I think that, for the array reader, accessing the value one past the end is an error. For `DICT`, we could simply declare that the value is NULL.

Then, I was going to sketch out the same behavior if we search for a key and found...nothing. Seems that the `DictReader` has no way to search for a key. If it did, it would not be clear what type to use for the key. So, that is a bit of a mess.

Let's assume we did have something to find a specific value, like we do for the arrays where we have `setPosn(int index)`. Maybe we have `findStringKey()` etc. For testing, we could just have `findKey(Object key)` and do a switch on type as we do for the other readers. So:

```
  dr.findStringKey("foo");
  if (vr.isNull()) {
    System.out.println("NULL");
  } else {
    System.out.println(vr.getAsString());
  }
```

Or, even:

```
  if (dr.findStringKey("foo")) {
    System.out.println(vr.getAsString());
  }
```

The only way to find the value is to do a linear search. If we do a linear search, and no value exists, the index will point one past the end of the dict. And, above, we suggested that the key and value readers could return `true` for `isNull()` in those cases.

Given this, unless I'm missing something, I don't see that we need any kind of null value reader at all. In fact, I'd even say that we don't want such a reader if we'd have to do:

```
ObjectReader kr = dr.findStringKey("foo");
if (kr.isNull) {
  // kr is a special null reader
} else {
  // kr is a Map reader
}
```

The above clutters the code and makes it impossible to cache the value reader, which will slow inner-most loops that want to use the reader abstractions.

What do you think?

Reply:
@paul-rogers, first of all, thank you for your thoughts - they are very useful.

Regarding the first case you described (I suppose you meant `ScalarReader kw` and `MapReader vr` in the first code snippet):
one is able to write code with cached readers currently, if needed for such iterations over key-value pairs in the form you wrote it
```
while (dr.next()) {
  System.out.println(kr.getString());
  System.out.println(vr.getAsString());
}
```
with the behaviour you expect (without a special `Null*Reader`s). But there is no need for the case (here and below) to add a special `NULL`-`NULL` entry past the end, as iteration should be over _true_ collection of entries, which were added to the `Dict` (through `DictWriter`).

Regarding finding specific values:
there is a `DictReader#getValueReader(Object key)` method (with return type `ObjectReader`), which allows to obtain `value`'s reader with its position set to the one corresponding to needed `key`. If there is no one for the `key` - the special `Null*Reader` (depends on the type of `value`) is returned. So, this approach enforces non-caching of the value reader (as you have pointed out). The idea was to make `null` `value`s to be handled the same, indepently whether the `null` is from the actual value (i.e., from some entry with some `key` and `null` `value` `"someKey" -> null`) or from absent `key`, e.g. your example:
```
ObjectReader kr = dr.findStringKey("foo");
if (kr.isNull) {
  // kr is a special null reader
} else {
  // kr is a Map reader
}
```
is likely should be done, because of the nature of complex types in Drill - they are generally non-nullable (at least the ones in standard, pre-EVF, Drill's framework) leading to no distinction between, say, empty or `null` `MAP`.

But good point about poorer perfomance (with the current approach).

I like your idea with `boolean findKey(...)` - something similar is done in standard framework as well. Perhaps, I will stick to it. (In this case, your suggestion about treating entries by the end as if they were `null` does make sense).

About designated method for each key type - I assume it is redundant, as `key` type validation is better to be performed once prior to obtaining the data, and general case of `Object key` should do, shouldn't it? 

(Just want to note, that array-like behaviour is useful in testing primarily, the "real" use-case is to obtain elements by key).

P.s. Sorry, missed the comment yesterday.

Reply:
Yes, there is a need to have multiple methods for each key type to avoid invocation of generalized `keyReader.getObject()` which contains `switch` checking each type (really bad in production) - will fix it.

Reply:
@KazydubB, first, thanks for understanding my hastily-written example; I've edited it to be accurate.

We are defining an API. APIs exist to support use cases, so it is very helpful to think about those. (My bad for not doing that earlier; I was focused more on the plumbing aspects early on. This is, in fact, why I still like design docs, even in the era of Agile, so we can visualize these issues early on.)

In my view, there are two key use cases: tests and production.

In tests, we simply want to verify that, say, a reader correctly created (or filtered, or altered) a `DICT`. If a key exists, we want to verify its value. If a key does not exist, we simply need to verify that fact. So, having a generic `boolean findKey(Object key)` is perfectly fine. The overhead of doing a type-based switch statement on every call is fine. (This is why tests often use the `addRow()` methods which, internally, have a big, slow `if` statement based on the type of the value.)

The second use case is production. Here, we can assume our goal is maximum performance via generated code.  I would argue that we do not want any extra code in the access path. Instead, the code should be specific to each type. This is why the readers are designed to go directly from, say, `getString()` into the code that grabs the bytes from the underlying value vector.

For production code, we'd want a type-aware "find" operation. There are a couple of ways to do that. One is by adding a "key accessor":

```
class DictReader ...
   KeyAccessor keyAccessor();
  ...
}

interface KeyAccessor {
  boolean findInt(int key);
  boolean findString(String key);
  ...
}
```

The accessor code gen could then generate type-specific search code by creating a key accessor class for each supported key type. This would be done in the build-time code gen, using Freemarker, as part of the existing `columnAccessors.java` template and code gen.

Now things get interesting. We need to generate code for each SQL operation that uses a `DICT`. Let's make up an example:

```
SELECT ... FROM myTableWithDict WHERE d['foo'] = 20
```

Here, the story splits into two paths. Today, code gen works directly against value vectors: you will need to generate the find code inline for each operator. Longer term, it is my hope that we can shift to using the column readers. (Doing so will simplify code gen, and puts us in a position to use multiple value vector format, such as Arrow, if we wanted.)

So, let's imagine we are using your `DICT` readers in runtime code generation. We want to generate code for the above example. 

In either case (old-school or reader-based), we don't know the type until the first row appears. At that point, we generate run-time code that, in our example, uses a `VARCHAR` key to lookup an `INT` value and send it to the equality function. Very crudely:

```
DictReader dr;
KeyAccessor ka;
ScalarReader vr;
String constKey;
String constArg;

void doSetup() {
  dr = rowReader.dict("d");
  ka = dr.keyAccessor();
  vr = dr.value().scalar();
  constKey = "foo";
  constArg = "bar";
}

void doEval() {
  boolean isEquals;
  boolean found = ka.findString(constKey);
  if (found) {
    isEquals = strEquals(vr.getString(), constArg);
  } else {
    isEquals = false;
  }
  // do something with isEquals...
}
```

The above is written in "code gen style", but greatly simplified.

So, we've looked at two use cases: test and production. In neither did we need to work with a missing value. Instead, we observed the rules that a non-existent key has no value (rather than a non-existent key having a value which is structurally the same as a real value, just null.)

Think of Java. A `null` in Java is the absence of a value, it is not, say a `Foo` object, with all the same fields and methods, that somehow has special `null` semantics. (Yes, this is a huge source of debate in CS. Some languages, like Scala, try to use objects, like `Option`, instead of `null`. But, `null` is still a higher-performance solution.) Further, in SQL, `NULL` is defined as the absence of a value, not a value (say a map) with special `NULL` semantics.

So, I don't think you need to try to define null readers: a `NULL` is missing value: there is nothing there, no structure to explore, no value to get.

Does this make sense?

In summary:

* Define the use cases you want to support.
* Define how those use cases will look up keys and how they will handle non-existent keys.
* Refine the reader API to work well for those use cases.

The good news is that the core functionality is now looking pretty good; we're just fine tuning the functionality to best serve the potential users of your new code.